test_conformance/extensions/cl_khr_kernel_clock/kernel_clock.cpp

```diff
@@ -0,0 +1,202 @@
+                   "Unable to query "
+                   "CL_DEVICE_KERNEL_CLOCK_CAPABILITIES_KHR");
+
+        // Skip if capability is not supported
+        return capability != (capabilities & capability);
+    }
+
+    cl_int RunTest()
+    {
+        size_t global_size = 1;
+        cl_uint buf = 0;
+        char kernel_src[512];
+        const char *ptr;
+        cl_int error;
+
+        // 2 built-ins for each scope
+        for (size_t i = 0; i < 2; i++)
+        {
+            buf = 0;
+            clProgramWrapper program;
+            clKernelWrapper kernel;
+            clMemWrapper out_mem;
+
+            if (i == 0 && !gHasLong)
+            {
+                log_info("The device does not support ulong. Testing hilo "
+                         "built-ins only\n");
+                continue;
+            }
+
+            switch (capability)
+            {
+                case CL_DEVICE_KERNEL_CLOCK_SCOPE_DEVICE_KHR: {
+                    sprintf(kernel_src, kernel_sources[i], "device", "device");
+                    break;
+                }
+                case CL_DEVICE_KERNEL_CLOCK_SCOPE_WORK_GROUP_KHR: {
+                    sprintf(kernel_src, kernel_sources[i], "workgroup",
+                            "workgroup");
+                    break;
+                }
+                case CL_DEVICE_KERNEL_CLOCK_SCOPE_SUB_GROUP_KHR: {
+                    sprintf(kernel_src, kernel_sources[i], "subgroup",
+                            "subgroup");
+                    break;
+                }
+            }
+
+            ptr = kernel_src;
+
+            error = create_single_kernel_helper_create_program(
+                context, &program, 1, &ptr);
+            test_error(error, "Failed to create program with source");
+
+            error =
+                clBuildProgram(program, 1, &device, nullptr, nullptr, nullptr);
+            test_error(error, "Failed to build program");
+
+            out_mem = clCreateBuffer(context, CL_MEM_WRITE_ONLY,
+                                     sizeof(cl_uint), nullptr, &error);
+            test_error(error, "clCreateBuffer failed");
+
+            kernel = clCreateKernel(program, "SampleClock", &error);
+            test_error(error, "Failed to create kernel");
+
+            error = clSetKernelArg(kernel, 0, sizeof(out_mem), &out_mem);
+            test_error(error, "clSetKernelArg failed");
+
+            error = clEnqueueNDRangeKernel(queue, kernel, 1, NULL, &global_size,
+                                           NULL, 0, NULL, NULL);
+            test_error(error, "clNDRangeKernel failed");
+
+            error = clEnqueueReadBuffer(queue, out_mem, CL_BLOCKING, 0,
+                                        sizeof(cl_uint), &buf, 0, NULL, NULL);
+            test_error(error, "clEnqueueReadBuffer failed");
+
+            if (buf == 1)
+            {
+                log_error(
+                    "Sampling the clock returned bad values, time1 > time2.\n");
```

Comment:
There was a bit of discussion on GitLab about this also.  I agree there could be a problem here in theory, but these tests are currently matching the behavior for the Vulkan tests, see:

https://github.com/KhronosGroup/VK-GL-CTS/blob/main/external/vulkancts/modules/vulkan/shaderexecutor/vktShaderClockTests.cpp#L182

Since there doesn't seem to have been a problem with the Vulkan tests in practice so far, I'm inclined to keep these tests as-is, and if we end up needing something fancier we can cross that bridge when we come to it.

Reply:
Thanks for cross-checking, that sounds reasonable to me then.